CIS 2.7.1 Error serverAddresses: Invalid type. Expected: array, given: null

#### Setup Details
 CIS Version : 2.7.1    
 Build: f5networks/k8s-bigip-ctlr:latest       
 BIGIP Version: Big IP 15.1.4
 AS3 Version: 3.20   
 Agent Mode: AS3     
 Orchestration: OSCP      
 Orchestration Version: 4.8.18
 Pool Mode: Cluster      
 Additional Setup details: OpenshiftSDN CNI

#### Description  
Hi after migrate from version 2.1.1 to 2.7.1  I observed errors in CIS logs:
2022/01/28 09:52:48 [ERROR] [AS3] Template is not valid. see errors
2022/01/28 09:52:48 [ERROR] - (root): Must validate one and only one schema (oneOf)
2022/01/28 09:52:48 [ERROR] - declaration: Must validate "then" as "if" was valid
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate "then" as "if" was valid
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate one and only one schema (oneOf)
2022/01/28 09:52:48 [ERROR] - declaration.members.0.serverAddresses: Invalid type. Expected: array, given: null
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration: Must validate "then" as "if" was valid
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate "then" as "if" was valid
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate one and only one schema (oneOf)
2022/01/28 09:52:48 [ERROR] - declaration.members.0.serverAddresses: Invalid type. Expected: array, given: null
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration: Must validate "then" as "if" was valid
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate "then" as "if" was valid
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate one and only one schema (oneOf)
2022/01/28 09:52:48 [ERROR] - declaration.members.0.serverAddresses: Invalid type. Expected: array, given: null
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/01/28 09:52:48 [ERROR] - declaration: Must validate all the schemas (allOf)

In version 2.1.1 this errors don't exists.
This is in situation, when in ConfigMap exists configuration for not existent Service object on Openshift and CIS can't find proper pods assigned to Service

Error leads to not sync configuration between F5 and OCSP.

#### Steps To Reproduce
1) Create config map with AS3 declaration with Service that not exists on OCSP
2) Look at the logs
3) Error should occur

#### Expected Result
CIS should sync configuration between F5 and OCSP

#### Actual Result  
Errors in CIS logs and not sync between F5 and OCSP

#### Observations (if any)

This is easly to reproduce.
In one of namespace I run scale replicas from 4 to 2 pods
Results:
When use version 2.7.1 -> on F5 pods are market as DOWN but not deleted from pool
When use version 2.1.1 -> on F5 pods are market as DOWN and after some time deleted from pool as expected.

Regards

>  configuration related to non existent Services, so create in that style is forrbiden ?

Yes. Without proper services created, EndPoints will not be created and CIS cannot generate correct AS3 with pool members. So you will see these errors. 

>2022/01/28 09:52:48 [ERROR] - declaration.members.0.serverAddresses: Invalid type. Expected: array, given: null

Hi
In logs I have:
2022/02/04 10:18:16 [DEBUG] [CORE] Finished syncing virtual servers example01-svc in namespace example01 (799.887443ms), 113/113
2022/02/04 10:18:16 [ERROR] [AS3] Template is not valid. see errors
2022/02/04 10:18:16 [ERROR] - (root): Must validate one and only one schema (oneOf)
2022/02/04 10:18:16 [ERROR] - declaration: Must validate "then" as "if" was valid
2022/02/04 10:18:16 [ERROR] - declaration.members.0: Must validate "then" as "if" was valid
2022/02/04 10:18:16 [ERROR] - declaration.members.0: Must validate one and only one schema (oneOf)
2022/02/04 10:18:16 [ERROR] - declaration.members.0.serverAddresses: Invalid type. Expected: array, given: null
2022/02/04 10:18:16 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/02/04 10:18:16 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/02/04 10:18:16 [ERROR] - declaration: Must validate all the schemas (allOf)
2022/02/04 10:18:49 [DEBUG] [CORE] Finished syncing virtual servers example02-svc in namespace example02 (793.883237ms), 113/113
2022/02/04 10:18:52 [ERROR] [AS3] Template is not valid. see errors
2022/02/04 10:18:52 [ERROR] - (root): Must validate one and only one schema (oneOf)
2022/02/04 10:18:52 [ERROR] - declaration: Must validate "then" as "if" was valid
2022/02/04 10:18:52 [ERROR] - declaration.members.0: Must validate "then" as "if" was valid
2022/02/04 10:18:52 [ERROR] - declaration.members.0: Must validate one and only one schema (oneOf)
2022/02/04 10:18:52 [ERROR] - declaration.members.0.serverAddresses: Invalid type. Expected: array, given: null
2022/02/04 10:18:52 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/02/04 10:18:52 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/02/04 10:18:52 [ERROR] - declaration: Must validate all the schemas (allOf)
2022/02/04 10:22:41 [DEBUG] [CORE] Finished syncing virtual servers example03 in namespace example03 (154.242µs), 113/113
2022/02/04 10:22:42 [ERROR] [AS3] Template is not valid. see errors
2022/02/04 10:22:42 [ERROR] - (root): Must validate one and only one schema (oneOf)
2022/02/04 10:22:42 [ERROR] - declaration: Must validate "then" as "if" was valid
2022/02/04 10:22:42 [ERROR] - declaration.members.0: Must validate "then" as "if" was valid
2022/02/04 10:22:42 [ERROR] - declaration.members.0: Must validate one and only one schema (oneOf)
2022/02/04 10:22:42 [ERROR] - declaration.members.0.serverAddresses: Invalid type. Expected: array, given: null
2022/02/04 10:22:42 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/02/04 10:22:42 [ERROR] - declaration.members.0: Must validate all the schemas (allOf)
2022/02/04 10:22:42 [ERROR] - declaration: Must validate all the schemas (allOf)

For which one tenant this ERRORS belongs ?
I see also in log:
2022/02/04 10:22:11 [DEBUG] [CORE] Discovered members for service b... is []
2022/02/04 10:22:11 [DEBUG] [CORE] Discovered members for service ... is []

But there is proper assigment but from some reasons pods not working for this service.

Regards

Hi
I know that is related to issue #1602, but if you have a lot of tenants and configmap, find in which one this occurs is difficult without information in log, you must check one by one and spend about this a lot of time.
Nicely will be for eexample the logs looks like:

2022/02/04 10:22:42 [ERROR] [AS3] Template is not valid. see errors (tenant: <name of namespace or configmap>)

Regards

--- a/pkg/agent/as3/as3ConfigMap.go
+++ b/pkg/agent/as3/as3ConfigMap.go
@@ -53,8 +53,8 @@ func (am *AS3Manager) prepareResourceAS3ConfigMaps() (
 			if am.as3Validation == true {
 				if ok := am.validateAS3Template(rscCfgMap.Data); !ok {
 					log.Errorf("[AS3][Configmap] Error validating AS3 template")
-					log.Errorf("[AS3][Configmap] Error in processing the ConfigMap: %v/%v",
-						rscCfgMap.Namespace, rscCfgMap.Name)
+					log.Errorf("[AS3][Configmap] Error in processing the resource ConfigMap: %v in Namespace: %v",
+						rscCfgMap.Name, rscCfgMap.Namespace)
 					continue
 				}
 			}

